modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java

```diff
@@ -343,4 +355,164 @@ static boolean hasSubQuery(SqlNode node) {
+            return null;
+        }
+
+        IgniteSelectCount rel = new IgniteSelectCount(
+                planner.cluster(),
+                planner.cluster().traitSetOf(IgniteConvention.INSTANCE),
+                targetTable,
+                expressions
+        );
+
+        return new Pair<>(rel, expressionNames);
+    }
+
+    private static @Nullable SqlSelect getSelectCountOptimizationNode(SqlNode node) {
+        // Unwrap SELECT .. from SELECT x FROM t ORDER BY ...
+        if (node instanceof SqlOrderBy) {
+            SqlOrderBy orderBy = (SqlOrderBy) node;
+
+            // Skip ORDER BY with OFFSET/FETCH
+            if (orderBy.fetch != null || orderBy.offset != null) {
+                return null;
+            }
+
+            // Skip ORDER BY with non literals
+            for (SqlNode arg : orderBy.orderList) {
+                if (!SqlUtil.isLiteral(arg))  {
+                    return null;
+                }
+            }
+
+            assert orderBy.getOperandList().size() == 4 : "Expected 4 operands, but was " + orderBy.getOperandList().size();
+            node = orderBy.query;
+        }
+
+        if (!(node instanceof SqlSelect)) {
+            return null;
+        }
+
+        SqlSelect select = (SqlSelect) node;
+
+        if (select.getGroup() != null
+                || select.getFrom() == null
+                || select.getWhere() != null
+                || select.getHaving() != null
+                || select.getQualify() != null
+                || !select.getWindowList().isEmpty()
+                || select.getOffset() != null
+                || select.getFetch() != null) {
+
+            return null;
+        }
+
+        // make sure that the following IF statement does not leave out any operand of the SELECT node
+        assert select.getOperandList().size() == 12 : "Expected 12 operands, but was " + select.getOperandList().size();
+
+        // Convert PUBLIC.T AS X (a,b) to PUBLIC.T
+        SqlNode from = SqlUtil.stripAs(select.getFrom());
+        // Skip non-references such as VALUES ..
+        if (from.getKind() == SqlKind.IDENTIFIER) {
+            return select;
+        } else {
+            return null;
+        }
+    }
+
+    private static boolean isCountStar(SqlNode node) {
+        if (!SqlUtil.isCallTo(node, SqlStdOperatorTable.COUNT)) {
+            // The SQL node was checked by the validator and the call has correct number of arguments.
+            return false;
+        } else {
+            SqlCall call = (SqlCall) node;
+            // Reject COUNT(DISTINCT ...)
+            if (call.getFunctionQuantifier() != null) {
+                return false;
+            }
+            if (call.getOperandList().isEmpty()) {
+                return false;
+            }
+            SqlNode operand = call.getOperandList().get(0);
+            return !isCountStartOperand(operand);
+        }
+    }
+
+    private static boolean isCountStartOperand(SqlNode node) {
```

Comment:
name of this method is quite confusing. Why `NULL` is countStart operand while any literal is not?

Reply:
Fixed. Moved the body of this method to `isCountStar`.